kong/dynamic_hook/init.lua

```diff
@@ -23,39 +29,123 @@ local non_function_hooks = {
 --]]
 }
 
-local always_enabled_groups = {}
-
-local wrap_functions = {
-  [0] = warp_function_gen.generate_wrap_function(0),
-  [1] = warp_function_gen.generate_wrap_function(1),
-  [2] = warp_function_gen.generate_wrap_function(2),
-  [3] = warp_function_gen.generate_wrap_function(3),
-  [4] = warp_function_gen.generate_wrap_function(4),
-  [5] = warp_function_gen.generate_wrap_function(5),
-  [6] = warp_function_gen.generate_wrap_function(6),
-  [7] = warp_function_gen.generate_wrap_function(7),
-  [8] = warp_function_gen.generate_wrap_function(8),
-  ["varargs"] = warp_function_gen.generate_wrap_function("varargs"),
-}
+
+local ALWAYS_ENABLED_GROUPS = {}
+
+
+local function should_execute_original_func(group_name)
+  if ALWAYS_ENABLED_GROUPS[group_name] then
+    return
+  end
+  
+  local phase = ngx_get_phase()
+  if phase == "init" or phase == "init_worker" then
+    return true
+  end
+
+  local dynamic_hook = ngx.ctx.dynamic_hook
+  if not dynamic_hook then
+    return true
+  end
+
+  local enabled_groups = dynamic_hook.enabled_groups
+  if not enabled_groups[group_name] then
+    return true
+  end
+end
+
+
+local function execute_hook_vararg(hook, hook_type, group_name, ...)
+  if not hook then
+    return
+  end  
+  local ok, err = pcall(hook, ...)
+  if not ok then
+    ngx_log(ngx_WARN, "failed to run ", hook_type, " hook of ", group_name, ": ", err)
+  end
+end
+
+
+local function execute_hooks_vararg(hooks, hook_type, group_name, ...)
+  if not hooks then
+    return
+  end
+  for _, hook in ipairs(hooks) do
+    execute_hook_vararg(hook, hook_type, group_name, ...)
+  end
+end
+
+
+local function execute_after_hooks_vararg(handlers, group_name, ...)
+  execute_hook_vararg(handlers.after_mut, "after_mut", group_name, ...)
+  execute_hooks_vararg(handlers.afters, "after", group_name, ...)
+  return ...
+end
+
+
+local function wrap_function_vararg(group_name, original_func, handlers)
+  return function (...)
+    if should_execute_original_func(group_name) then
+      return original_func(...)
+    end
+    execute_hooks_vararg(handlers.befores, "before", group_name, ...)
+    return execute_after_hooks_vararg(handlers, group_name, original_func(...))
+  end
+end
+
+
+local function execute_hook(hook, hook_type, group_name, a1, a2, a3, a4, a5, a6, a7, a8)
```

Comment:
Some functions (such as cosocket) has strict checks of the number of arguments, so we cannot do something like `a0, a1, a2, a3 ......`.

Reply:
Perhaps this problem should be handled on hook side.

Reply:
I made a workaround to C code thing.

Reply:
> @ADD-SP you mean those that are implemented in C? (in Lua land it most likely doesn't matter)

Yes, the Lua C function might have strict checks on the number of arguments.